src/csharp/Microsoft.Spark/Utils/UdfSerDe.cs

```diff
@@ -68,7 +71,50 @@ internal sealed class FieldData
         {
             public TypeData TypeData { get; set; }
             public string Name { get; set; }
+            public ValueData ValueData { get; set; }
+        }
+
+        [Serializable]
+        internal sealed class ValueData : ISerializable
+        {
+            public ValueData() { }
+
+            public TypeData TypeData { get; set; }
+
             public object Value { get; set; }
+
+            public void GetObjectData(SerializationInfo info, StreamingContext context)
+            {
+                info.AddValue("TypeData", TypeData, typeof(TypeData));
+
+                var bf = new BinaryFormatter();
+                using (var ms = new MemoryStream())
+                {
+                    bf.Serialize(ms, Value);
+                    info.AddValue("ValueSerialized", ms.ToArray(), typeof(byte[]));
+                }
+            }
+
+            public ValueData(SerializationInfo info, StreamingContext context)
+            {
+                TypeData = (TypeData)info.GetValue("TypeData", typeof(TypeData));
+
+                var valueSerialized = (byte[])info.GetValue("ValueSerialized", typeof(byte[]));
+                var bf = new BinaryFormatter();
+                using (var ms = new MemoryStream(valueSerialized, false))
+                {
+                    try
+                    {
+                        Value = bf.Deserialize(ms);
+                    }
+                    catch (SerializationException)
+                    {
+                        ms.Seek(0, SeekOrigin.Begin);
+                        DeserializeType(TypeData);
+                        Value = bf.Deserialize(ms);
+                    }
+                }
+            }
         }
 
         internal static UdfData Serialize(Delegate udf)
```

Comment:
Can you add few tests to Microsoft.Spark.UnitTest for UdfSerDe class? (And an example with TargetData - it doesn't have to an external dll).

Reply:
Added some test cases.